qualtran/bloqs/chemistry/prepare_mps/synthesize_gate_hr.py

```diff
@@ -0,0 +1,237 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Dict, List, Sequence, Tuple
+
+import attrs
+import numpy as np
+
+from qualtran import Bloq, BloqBuilder, BoundedQUInt, Register, Signature, SoquetT
+from qualtran.bloqs.basic_gates import CNOT, Hadamard, ZGate
+from qualtran.bloqs.basic_gates.x_basis import XGate
+from qualtran.bloqs.reflection_using_prepare import ReflectionUsingPrepare
+from qualtran.bloqs.rotations.phase_gradient import PhaseGradientState
+from qualtran.bloqs.select_and_prepare import PrepareOracle
+from qualtran.bloqs.state_preparation.state_preparation_via_rotation import (
+    StatePreparationViaRotations,
+)
+
+
+@attrs.frozen
+class SynthesizeGateViaHR(Bloq):
+    r"""Generates an unitary via Householder reflections given all or some of its columns.
+
+    This bloq implements the algorithm described in [1], which basically implements an arbitrary
+    unitary operator $U$ with a reflection ancilla as
+
+    $$
+        |0\rangle\langle 1| U + |1\rangle\langle 0| U^\dagger$,
+    $$
+
+    even though for the purpose of making this bloq simpler to work with, the gate actually
+    implemented is
+
+    $$
+        |0\rangle\langle 1| U + |1\rangle\langle 0| U^\dagger$.
```

Comment:
Yes, sorry. The second one should have been $|0\rangle\langle 0| U + |1\rangle\langle 1| U^\dagger$, that is, instead of flipping the ancilla it just takes it and gives it back both as |0>. This makes using the block easier because when freeing the ancilla it is projected to zero.